### PR TITLE
detect-pcre-v1: add tests to detect pcre

### PR DESCRIPTION
**Pull request by TheKharleeci**
detect-pcre-01: DetectPcreModifPTest04
detect-pcre-02: DetectPcreModifPTest05
detect-pcre-03: DetectPcreTestSig01-03
detect-pcre-04: DetectPcreTestSig09-14
detect-pcre-05: DetectPcreTestSig15-16

### CODE REVIEW

**Comment by TheKharleeci** [comment]
These are the conversions I have done on detect-pcre. There are some other tests I am not sure that I should convert

File: tests/writepcap.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env python
+from scapy.all import *
+
+pkts = []
+
+pkts += Ether(dst='ff:ff:ff:ff:ff:ff', src='00:01:02:03:04:05')/ \
+    Dot1Q(vlan=6)/ \
+    IP(dst='1.2.3.4', src='5.6.7.8')/TCP(sport=6666, dport=80, flags='P''A')/"POST / HTTP/1.0\r\nUser-Agent: Mozilla/1.0\r\nCookie: dummy 1234\r\n\r\n"
+wrpcap('input-pcre-05.pcap', pkts)
```
Comment (jufajardini) [inline tests/writepcap.py:9]: I think this was left behind :P

Comment (TheKharleeci) [reply tests/writepcap.py:9]: Oops. WIll remove it. used it for to test

**Comment by jufajardini** [comment]
> These are the conversions I have done on detect-pcre. There are some other tests I am not sure that I should convert

DetectPcreParseTest01 to DetectPcreParseTest28 don't seem like good candidates to the type of conversion you're doing, as they're for signature parsing, from what I can see.

Checking the rest :)

**Review by jufajardini — COMMENTED** [review]
Thanks for this work, added a few comments. :)

File: tests/detect-pcre/detect-pcre-03/test.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+      event_type: flow
+- filter:
+    count: 1
+    match:
+      event_type: stats
+- filter:
+    count: 1
+    match:
+      event_type: alert
+      alert.signature_id: 2
```
[the remark this answers is not in the document]
Comment (TheKharleeci) [reply tests/detect-pcre/detect-pcre-03/test.yaml:17]: Ok. I'll be adding that

File: tests/detect-pcre/detect-pcre-02/test.rules [hunk 3]
```diff
@@ -0,0 +1,2 @@
+alert http any any -> any any (msg:"Pcre modifier P"; pcre:"/DOC/P"; sid:1;)
+alert http any any -> any any (msg:"Pcre modifier P (no match)"; pcre:"/DOCTYPE/P"; sid:2;)
```
[the remark this answers is not in the document]
Comment (inashivb) [reply tests/detect-pcre/detect-pcre-02/test.rules:2]: I think it is exactly the opposite. Let's try to evaluate though. We all know how easy binary is. :P
`FAIL_IF(PacketAlertCheck(p1, 1));`
`FAIL_IF(true) -> FAIL`
`FAIL_IF(false) -> CONTINUE`

so, `FAIL_IF(PACKET_ALERTS ? true : false) = FAIL_IF(Packet DOES Alert)`

Now, I may be very wrong for the reason mentioned above. Binary is SO easy and not at all confusing at times. But, WDYT?

Anyhow, I think this test does need fixing. I'll get back in a while about how.

[the remark this answers is not in the document]
Comment (jufajardini) [reply tests/detect-pcre/detect-pcre-02/test.rules:2]: My bad, yes, I read that part totally wrong xD But what I meant was that from the checks in the test, it seems that for `p1`, both rules shouldn't alert (as you said), whereas they should be triggered with `p2`. But rule with sid:2 has "(no match)" in the message, so it got me thinking if the original intention was for it not to match? I'm totally ignorant...

[the remark this answers is not in the document]
Comment (inashivb) [reply tests/detect-pcre/detect-pcre-02/test.rules:2]: I think it was copy pasted from the test before this one and the message was not updated.

And, yes this test is incorrect. Getting back in some time to review this PR.

File: tests/detect-pcre/detect-pcre-02/writepcap.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+from scapy.all import *
+
+pkts = []
+
+pkts += Ether(dst='ff:ff:ff:ff:ff:ff', src='00:01:02:03:04:05')/ \
+    Dot1Q(vlan=6)/ \
+    IP(dst='1.2.3.4', src='5.6.7.8')/TCP(sport=6666, dport=80, flags='P''A')/"GET / HTTP/1.1\r\nHost: www.emergingthreats.net\r\nUser-Agent: Mozilla/1.0\r\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9;q=0.8\r\nAccept-Encoding: gzip,deflate\r\nTransfer-Encoding: chunked\r\nContent-Type: text/html; charset=utf-8\r\n\r\n15\r\n<!DOC"
+pkts += Ether(dst='ff:ff:ff:ff:ff:ff', src='00:01:02:03:04:05')/ \
+    Dot1Q(vlan=6)/ \
+    IP(dst='1.2.3.4', src='5.6.7.8')/TCP(sport=6666, dport=90, flags='P''A')/"<!DOCTYPE html PUBLIC\r\n0\r\n\r\n"
```
Comment (victorjulien) [inline tests/detect-pcre/detect-pcre-02/writepcap.py:11]: the previous payload ends with `<!DOC`, but here we have `<!DOC` again, is that intended?

File: tests/detect-pcre/detect-pcre-03/test.rules [hunk 5]
```diff
@@ -0,0 +1,3 @@
+alert tcp any any -> any any (msg:"HTTP TEST"; pcre:"/^gEt/i"; pcre:"/\\/two\\//U"; pcre:"/GET \\/two\\//"; pcre:"/\\s+HTTP/R"; sid:1;)
```
Comment (inashivb) [inline tests/detect-pcre/detect-pcre-03/test.rules:1]: hmm I see in the C tests that this should have alerted?
```
FAIL_IF_NOT(PacketAlertCheck(p, 1) == 1);
```
I'm wondering if it has to do with the regex and escape sequences. Don't know either well. @catenacyber could you please check if this rule is correct?

File: tests/detect-pcre/detect-pcre-04/test.rules [hunk 6]
```diff
@@ -0,0 +1,3 @@
+alert http any any -> any any (msg:"HTTP cookie"; pcre:"/dummy/C"; sid:1;)
+alert http any any -> any any (msg:"HTTP cookie"; pcre:!"/dummy/C"; sid:2;)
+alert http any any -> any any (msg:"HTTP method"; pcre:"/POST/M"; sid:3;)
```
[the remark this answers is not in the document]
Comment (inashivb) [reply tests/detect-pcre/detect-pcre-04/test.rules:3]: Agreed w Juliana. Please add all these rules and corresponding filters too.

**Review by inashivb — CHANGES_REQUESTED** [review]

File: tests/detect-pcre/detect-pcre-03/test.rules (hunk 5 above)
Comment (inashivb) [reply tests/detect-pcre/detect-pcre-03/test.rules:1]: Figured it out w the help from Victor, the escape sequence is incorrect. `\\` is C's escape sequence, so, if you change the rule to 
```
alert tcp any any -> any any (msg:"HTTP TEST"; pcre:"/^gEt/i"; pcre:"/\/two\//U"; pcre:"/GET \/two\//"; pcre:"/\s+HTTP/R"; sid:1;)
```
it works as expected :)

File: tests/detect-pcre/detect-pcre-02/writepcap.py (hunk 4 above)
Comment (catenacyber) [reply tests/detect-pcre/detect-pcre-02/writepcap.py:11]: I think it was indeed like this in C, but not intended as the comment states `DOCTYPE fragmented`

File: tests/detect-pcre/detect-pcre-02/test.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+args:
+- --set stream.midstream=true
+
+checks:
+- filter:
+    count: 2
+    match:
+      event_type: flow
```
[the remark this answers is not in the document]
Comment (TheKharleeci) [reply tests/detect-pcre/detect-pcre-02/test.yaml:8]: Ok. Thank you for pointing it out

File: tests/detect-pcre/detect-pcre-02/writepcap.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+wrpcap('input-pcre-05.pcap', pkts)
```
[the remark this answers is not in the document]
Comment (TheKharleeci) [reply tests/detect-pcre/detect-pcre-02/writepcap.py:12]: This was an error on my part. would be corrected in the next PR

File: tests/detect-pcre/detect-pcre-03/test.rules (hunk 5 above)
Comment (TheKharleeci) [reply tests/detect-pcre/detect-pcre-03/test.rules:1]: Ok. Thank you

**Comment by TheKharleeci** [comment]
Followed By: #717

**Pull request closed without merging**